plugins/outputs/syslog/syslog_mapper.go

```diff
@@ -0,0 +1,201 @@
+		sm.mapStructuredDataItem(tag.Key, tag.Value, msg)
+	}
+	for _, field := range metric.FieldList() {
+		sm.mapStructuredDataItem(field.Key, formatValue(field.Value), msg)
+	}
+}
+
+func (sm *SyslogMapper) mapStructuredDataItem(key string, value string, msg *rfc5424.SyslogMessage) {
+	if sm.reservedKeys[key] {
+		return
+	}
+	isExplicitSdid := false
+	for _, sdid := range sm.Sdids {
+		k := strings.TrimLeft(key, sdid+sm.Separator)
+		if len(key) > len(k) {
+			isExplicitSdid = true
+			msg.SetParameter(sdid, k, value)
+			break
+		}
+	}
+	if !isExplicitSdid && len(sm.DefaultSdid) > 0 {
+		k := strings.TrimPrefix(key, sm.DefaultSdid+sm.Separator)
+		msg.SetParameter(sm.DefaultSdid, k, value)
+	}
+}
+
+func (sm *SyslogMapper) mapAppname(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetTag("appname"); ok {
+		msg.SetAppname(formatValue(value))
+	} else {
+		//Use default appname
+		msg.SetAppname(sm.DefaultAppname)
+	}
+}
+
+func mapMsgID(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetField("msgid"); ok {
+		msg.SetMsgID(formatValue(value))
+	} else {
+		// We default to metric name
+		msg.SetMsgID(metric.Name())
+	}
+}
+
+func mapVersion(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetField("version"); ok {
+		if v, err := parseIntValue(value); err == nil {
+			msg.SetVersion(uint16(v))
+			return
+		}
+	}
+	msg.SetVersion(1)
+}
+
+func mapMsg(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetField("msg"); ok {
+		msg.SetMessage(formatValue(value))
+	}
+}
+
+func mapProcID(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetField("procid"); ok {
+		msg.SetProcID(formatValue(value))
+	}
+}
+
+func (sm *SyslogMapper) mapPriority(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	severityCode := sm.DefaultSeverityCode
+	facilityCode := sm.DefaultFacilityCode
+
+	if value, ok := getFieldCode(metric, "severity_code"); ok {
+		severityCode = *value
+	}
+
+	if value, ok := getFieldCode(metric, "facility_code"); ok {
+		facilityCode = *value
+	}
+
+	priority := (8 * facilityCode) + severityCode
+	msg.SetPriority(priority)
+}
+
+func mapHostname(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	// Try with hostname, then with source, then with host tags, then take OS Hostname
+	if value, ok := metric.GetTag("hostname"); ok {
+		msg.SetHostname(formatValue(value))
+	} else if value, ok := metric.GetTag("source"); ok {
+		msg.SetHostname(formatValue(value))
+	} else if value, ok := metric.GetTag("host"); ok {
+		msg.SetHostname(formatValue(value))
+	} else if value, err := os.Hostname(); err == nil {
+		msg.SetHostname(value)
+	}
+}
+
+func mapTimestamp(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	timestamp := metric.Time()
+	if value, ok := metric.GetField("timestamp"); ok {
+		if v, err := parseIntValue(value); err == nil {
+			timestamp = time.Unix(v, 0).UTC()
```

Comment:
Does this need to be `time.Unix(0, v)`?

Reply:
I forgot to address this remark from the other PR, sorry. Will do that soon

Reply:
If I remove the UTC() call, then I the timestamp is written with time offset.

`-<26>2 2010-11-10T23:30:00Z testhost testapp 25 555 - Test message`
`+<26>2 2010-11-11T00:30:00+01:00 testhost testapp 25 555 - Test message`

That's why I'd added the UTC() call in the first place.

Reply:
I'm actually referring to the order of the arguments to Unix(s, ns).  I looked at the syslog input and it appears that it writes the timestamp in nanoseconds so don't we need to pass it back in as nanoseconds here in order to roundtrip correctly?

Reply:
Yep, that's already fixed in the last commit:
https://github.com/influxdata/telegraf/blob/004fd1f924d50c15843d5b72d78afb32e3709b89/plugins/outputs/syslog/syslog_mapper.go#L148

but I kept the trailing UTC() function, otherwise the time offset is added into the syslog message.